crypt-gost-yescrypt.c

```diff
@@ -0,0 +1,182 @@
+ * modification, are permitted.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE AUTHOR AND CONTRIBUTORS ``AS IS'' AND
+ * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ * ARE DISCLAIMED.  IN NO EVENT SHALL THE AUTHOR OR CONTRIBUTORS BE LIABLE
+ * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+ * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
+ * OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
+ * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
+ * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
+ * SUCH DAMAGE.
+ */
+
+#include "crypt-port.h"
+
+#if INCLUDE_gost_yescrypt
+
+#include "crypt-private.h"
+#include "byteorder.h"
+
+#include <errno.h>
+
+#define YESCRYPT_INTERNAL
+#include "alg-yescrypt.h"
+#undef YESCRYPT_INTERNAL
+
+#include "alg-gost3411-2012-hmac.h"
+
+/* upper level hmac for tests */
+#ifndef outer_gost_hmac256
+#define outer_gost_hmac256 gost_hmac256
+#endif
+
+
+/* For use in scratch space by crypt_gost_yescrypt_rn().  */
+typedef struct
+{
+  yescrypt_local_t local;
+  gost_hmac_256_t gostbuf;
+  uint8_t outbuf[CRYPT_OUTPUT_SIZE];
+  uint8_t gsetting[CRYPT_OUTPUT_SIZE];
+  uint8_t hk[32], interm[32], y[32];
+  uint8_t *retval;
+} crypt_gost_yescrypt_internal_t;
+
+static_assert (sizeof (crypt_gost_yescrypt_internal_t) <= ALG_SPECIFIC_SIZE,
+               "ALG_SPECIFIC_SIZE is too small for GOST-YESCRYPT.");
+
+/*
+ * As OUTPUT is initialized with a failure token before gensalt_yescrypt_rn
+ * is called, in case of an error we could just set an appropriate errno
+ * and return.
+ */
+void
+gensalt_gost_yescrypt_rn (unsigned long count,
+                          const uint8_t *rbytes, size_t nrbytes,
+                          uint8_t *output, size_t o_size)
+{
+  if (o_size < 4 + 8 * 6 + BASE64_LEN (nrbytes) + 1 ||
+      CRYPT_GENSALT_OUTPUT_SIZE < 4 + 8 * 6 + BASE64_LEN (nrbytes) + 1)
+    {
+      errno = ERANGE;
+      return;
+    }
+
+  /* We pass 'o_size - 1' to gensalt, because we need to shift
+           the prefix by 1 char to insert the gost marker.  */
+  gensalt_yescrypt_rn (count, rbytes, nrbytes, output, o_size - 1);
+
+  /* Check for failures.  */
+  if (output[0] == '*')
+    return;
+
+  /* Shift output one byte further.  */
+  memmove (output + 1, output, strlen ((const char *) output) + 1);
+
+  /* Insert the gost marker.  */
+  output[1] = 'g';
+}
+
+void
+crypt_gost_yescrypt_rn (const char *phrase, size_t phr_size,
+                        const char *setting, size_t set_size,
+                        uint8_t *output, size_t o_size,
+                        void *scratch, size_t s_size)
+{
+  if (o_size < set_size + 1 + 43 + 1 ||
+      CRYPT_OUTPUT_SIZE < set_size + 1 + 43 + 1 ||
```

Comment:
That is not fully true:

* It may be the same size as `set_size`, if you want to verify `phrase` against `setting`.
* It may be `set_size + 1 + 43 + 1`, if you generate a new hash for `phrase` and `setting` is the pure output of `crypt_gensalt()`.

So we play it safe here and calculate for the second case.